### ISSUE TITLE
Getting to work with MultiProblem

### ISSUE THREAD

**Issue by stefan-falk** [tensorflow/tensor2tensor#1659]
#### Please refer to https://github.com/tensorflow/tensor2tensor/issues/1687

----

For training models, I have separated the data generation pipeline from t2t. For that I have implemented my own problem which, in essence, already expects a created dataset.

```python
@registry.register_problem
class ConfigBasedTranslationProblem(translate.TranslateProblem):

    # ...

    def training_filepaths(self, data_dir, num_shards, shuffled):
        return glob.glob(os.path.join(data_dir, '*.train.shard'))

    # ...

    def filepattern(self, data_dir, split: str, shard=None):
        split = 'dev' if split == 'eval' else split
        return os.path.join(data_dir, '*.%s.shard' % split)

    def generate_data(self, data_dir, tmp_dir, task_id=-1):
        raise NotImplementedError('Data should already be generated.')

    def prepare_to_generate(self, data_dir, tmp_dir):
        raise NotImplementedError('Data should already be generated.')

    def source_data_files(self, dataset_split):
        raise NotImplementedError('Data should already be generated.')
```

This works great so far but now that I discovered `MultiProblem` ([`multi_problem.md`](https://github.com/tensorflow/tensor2tensor/blob/master/docs/multi_problem.md)) I am facing some issues and questions I was hoping to be able to clarify here.

### The Language Model in `MultiProblem` (?)

This is not really a question but more a suggestion for code and API changes. I would provide a PR but atm I am stuck with tensor2tensor 1.12 and until I updated to 1.13 I won't have resources for it.

From  ([`multi_problem.md`](https://github.com/tensorflow/tensor2tensor/blob/master/docs/multi_problem.md)) and the code I can see that the first `problem` has to be a language-model problem.

From the looks, this seems like a requirement which could be relaxed. The first task seems to be used just to create a vocabulary for all languages. In my implementation, I just merge all vocabularies from all tasks (here called `dataset`) to one and return a `SubwordEncoder`:

```python
datasets = self.get_datasets()
reserved_tokens = set()
subword_tokens = set()
for dataset in datasets:
    encoders = dataset.feature_encoders()
    for encoder in encoders.values():
        encoder = cast(SubwordEncoder, encoder)
        reserved_tokens.update(encoder.reserved_tokens)
        subword_tokens.update(encoder.subword_tokens)

final_subwords = list(reserved_tokens) + sorted(list(subword_tokens.difference(reserved_tokens)))
# ...
return SubwordEncoder(vocab_fp)
``` 

The only other occurrence of the primary task I can see is in `get_hparams()` in order to set the vocab size and modality.

Imho it could make sense to relax this requirement if `MultiProblem` itself provided feature encoders for inputs and targets. All that is required for this would be additional "merge-vocab" logic and `MultiProblem` could work without the first task having to be a language model problem.

### How does `MultiProblem` training work?

In the `MultiProblem` class we find this:

```python
task_dataset = task_dataset.map(
  lambda x: self.add_task_id(task, x, enc, hparams, is_infer))
```
and `add_task_id()` takes an `example` (here `x`) in order to create

```python
# not is_infer 
inputs = example.pop("inputs")
concat_list = [inputs, [task.task_id], example["targets"]]
example["targets"] = tf.concat(concat_list, axis=0)
```
in case the problem has inputs or

```python
concat_list = [[task.task_id], example["targets"]]
example["targets"] = tf.concat(concat_list, axis=0)
```

in case the problem has no inputs.

Now, I do not quite understand why a `MultiProblem` only works on examples which provide `targets`. I was under the impression that we still simply train a `Transformer` model which gets `inputs` and `targets` presented as usual but that these samples are drawn from a set of tasks (problems/corpora) for training (see `multiproblem_per_task_threshold`).

So in theory, I should be able to create a classic `Text2TextProblem` which contains all required samples (plus a `task_id` for each sample) and it _should_ work as well, right? Or does `MultiProblem` work totally different?

The `task_id` does not set a different "_mode_" or something, it's just additional context for the model, isn't it?

### The Transformer in `MultiProblem`

I noticed that the graph of the Transformer looks very different from why I know from single-problem training. In particular, I noticed that there are no encoder layers? The `body` only contains the `decoder` layers. Why is this the case or what am I missing here?

> *Note:* It should not matter but this is from the `EvolvedTransformer` in particular

![image](https://user-images.githubusercontent.com/43335432/62944531-2bb1b780-bddd-11e9-89a0-41aa84e10ef4.png)

Does this mean hparams like `num_encoder_layers` are getting ignored here?

### `MultiProblem` Training Best Practice

I am not at the point where I could make use of recommendations but what I was thinking of was something like the following:

I want to test if (or how much) a small dataset can benefit from `MultiProblem` training. For this I would like to use a small `de_en` dataset. The main task of the `MultiProblem` would be to learn `en2de`. In addition I would like to use a `en_fr`. So the second task would be  a training on `en2fr` translation.

Setting `multiproblem_per_task_threshold` to`"95,5"` would mean that batches should consist of 95% `en2de` and 5% `en2fr` samples, is that correct? If so, can I expect improvements for the `"constant"` schedule or should I rather consider the `"pretrain"` schedule?

Any comments on that would be appreciated.

----

There are some other things I do not understand. Some which are inside the code of  `MultiProblem` e.g. the following:

```python
def dataset(self, ...): 
     # ..
      if not is_training and not is_infer:
        zeros = tf.zeros([self._ADDED_EVAL_COUNT, 1], dtype=tf.int64)
        pad_data = tf.data.Dataset.from_tensor_slices({
            "targets": zeros,
            "batch_prediction_key": zeros,
            "task_id": zeros,
        })
        task_dataset = task_dataset.concatenate(pad_data)
    # ..
```

which I 1) do not understand why it's there and 2) breaks my pipeline. I am overriding `example_reading_spec()` in my problems because I am adding things like the `corpus`-name which I use during evaluation. 

```python
def example_reading_spec(self):
    data_fields = {
        'targets': tf.VarLenFeature(tf.int64),
        'corpus': tf.VarLenFeature(tf.int64)
    }
    data_items_to_decoders = None
    return data_fields, data_items_to_decoders
```

Since this padding above is hard-coded, the program crashes. I had to override `MultiProblem.dataset` and add a dummy. What is this padding good for?

----

I know this is a lot so thank you for any response to these questions.
